tensorboard/plugins/graph/keras_util.py

```diff
@@ -0,0 +1,225 @@
+
+They have distinct structures to the configurations in shapes below:
+Functional:
+  config
+    name: Name of the model. If not specified, it is 'model' with
+          an optional suffix if there are more than one instance.
+    input_layers: Keras.layers.Inputs in the model.
+    output_layers: Layer names that are outputs of the model.
+    layers: list of layer configurations.
+      layer: [*]
+        inbound_nodes: inputs to this layer.
+
+Sequential:
+  config
+    name: Name of the model. If not specified, it is 'sequential' with
+          an optional suffix if there are more than one instance.
+    layers: list of layer configurations.
+      layer: [*]
+
+[*]: Note that a model can be a layer.
+Please refer to https://github.com/tensorflow/tfjs-layers/blob/master/src/keras_format/model_serialization.ts
+for more complete definition.
+"""
+from tensorboard.compat.proto.graph_pb2 import GraphDef
+from tensorboard.compat.tensorflow_stub import dtypes
+
+
+def _walk_layers(keras_layer):
+  """Walks the nested keras layer configuration in preorder.
+  Args:
+    keras_layer: Keras configuration from model.to_json.
+
+  Yields:
+    (name_scope, layer_config)
+  """
+  yield ('', keras_layer)
+  if keras_layer.get('config').get('layers'):
+    name_scope = keras_layer.get('config').get('name')
+    for layer in keras_layer.get('config').get('layers'):
+      for (sub_name_scope, sublayer) in _walk_layers(layer):
+        sub_name_scope = '%s/%s' % (
+            name_scope, sub_name_scope) if sub_name_scope else name_scope
+        yield (sub_name_scope, sublayer)
+
+
+def _scoped_name(name_scope, node_name):
+  """Returns scoped name for a node as a string in the form '<scope>/<node name>'.
+
+  Args:
+    name_scope: scope names similar to that of tf.name_scope.
+    node_name: current node name.
+
+  Returns
+    A scoped name.
+  """
+  if name_scope:
+    return '%s/%s' % (name_scope, node_name)
+  return node_name
+
+
+def _is_model(layer):
+  """Returns True if layer is a model."""
+  return layer.get('config').get('layers') is not None
+
+
+def _norm_to_list_of_layers(maybe_layers):
+  """Normalizes to a list of layers.
+
+  Args:
+    maybe_layers: A list of data[1] or a list of list of data.
+
+  Returns:
+    List of list of data.
+
+  [1]: A Functional model has fields 'inbound_nodes' and 'output_layers' which can
+  look like below:
+  - ['in_layer_name', 0, 0]
+  - [['in_layer_is_model', 1, 0], ['in_layer_is_model', 1, 1]]
+  The data inside the list seems to describe [name, size, index].
+  """
+  return (maybe_layers if isinstance(maybe_layers[0], (list,))
+          else [maybe_layers])
+
+
+def _update_dicts(name_scope,
+                  model_layer,
+                  input_to_in_layer,
+                  model_name_to_output,
+                  prev_node_name):
+  """Updates input_to_in_layer, model_name_to_output, and prev_node_name
+  based on the model_layer.
+
+  Args:
+    name_scope: Name scope of a model
```

Comment:
Please document argument types, i.e.,

name_scope: string representing name scope of a model
input_to_in_layer: dict of Keras.layers.input to inbound layer

etc.